File: lab/python/somas/soma_matrizes.py
def soma_matrizes(matriz1, matriz2):
    if len(matriz1) != len(matriz2):
        raise IndexError("MATRIZES PRECISAM TER O MESMO TAMANHO")

    nova_matriz = []
    for i in range(len(matriz1)):
        lista = []
        for j in range(len(matriz2)):
            lista.append(matriz1[i][j] + matriz2[i][j])
        nova_matriz.append(lista)

    return nova_matriz


def soma_matrizes1(matriz1, matriz2):
    if len(matriz1) != len(matriz2):
        raise IndexError("MATRIZES PRECISAM TER O MESMO TAMANHO")
    for i in range(len(matriz1)):
        for j in range(len(matriz2)):
            matriz1[i][j] = matriz1[i][j] + matriz2[i][j]

    return matriz1


def soma_matrizes2(matriz1, matriz2):
    return [
        [
            matriz1[i][j] + matriz2[i][j]
            for j in range(len(matriz1[i]))
        ]
        for i in range(len(matriz1))
    ]


def soma_matrizes3(matriz1, matriz2):
    return [
        [
            x + y
            for x, y in zip(l1, l2)
        ]
        for l1, l2 in zip(matriz1, matriz2)
    ]


print(soma_matrizes(matriz1=[[1, 2], [3, 4]], matriz2=[[5, 6], [7, 8]]))
print(soma_matrizes1(matriz1=[[1, 2], [3, 4]], matriz2=[[5, 6], [7, 8]]))
print(soma_matrizes2(matriz1=[[1, 2], [3, 4]], matriz2=[[5, 6], [7, 8]]))
print(soma_matrizes3(matriz1=[[1, 2], [3, 4]], matriz2=[[5, 6], [7, 8]]))

File: lab/c/somas/soma_matrizes.c
#include <stdio.h>


void soma_matrizes(int linhas, int colunas, int matriz1[linhas][colunas], int matriz2[linhas][colunas]){
    int nova_matriz[linhas][colunas];

    for (int i=0; i < linhas; i++){
        for (int j=0; j < colunas; j++){
            nova_matriz[i][j] = matriz1[i][j] + matriz2[i][j]; 
        }
    }

    // impressão bonitinha
    printf("[");
    for (int i = 0; i < linhas; i++) {
        printf("[");
        for (int j = 0; j < colunas; j++) {
            printf("%d", nova_matriz[i][j]);
            if (j < colunas - 1) printf(", ");
        }
        printf("]");
        if (i < linhas - 1) printf(", ");
    }
    printf("]\n");
}

void soma_matrizes1(int linhas, int colunas, int matriz1[linhas][colunas], int matriz2[linhas][colunas]){
    for (int i=0; i < linhas; i++){
        for (int j=0; j < colunas; j++){
            matriz1[i][j] = matriz1[i][j] + matriz2[i][j]; 
        }
    }

    // impressão bonitinha
    printf("[");
    for (int i = 0; i < linhas; i++) {
        printf("[");
        for (int j = 0; j < colunas; j++) {
            printf("%d", matriz1[i][j]);
            if (j < colunas - 1) printf(", ");
        }
        printf("]");
        if (i < linhas - 1) printf(", ");
    }
    printf("]\n");
}


int main() {
    int m1[2][2] = {{1, 2}, {3, 4}};
    int m2[2][2] = {{5, 6}, {7, 8}};

    soma_matrizes(2, 2, m1, m2);
    soma_matrizes1(2, 2, m1, m2);

    return 0;
}
